src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,218 @@
+            _functionInvocationDispatcherFactory = functionInvocationDispatcherFactory ?? throw new ArgumentNullException(nameof(functionInvocationDispatcherFactory));
+
+            _logger = loggerFactory?.CreateLogger<WorkerConcurrencyManager>();
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (_concurrencyOptions.Value.Enabled)
+            {
+                // Delay monitoring
+                await Task.Delay(_concurrencyOptions.Value.AdjustmentPeriod);
+
+                _functionInvocationDispatcher = _functionInvocationDispatcherFactory.GetFunctionDispatcher();
+                if (_functionInvocationDispatcher is HttpFunctionInvocationDispatcher)
+                {
+                    _logger.LogDebug($"Http worker concurrency is not supported.");
+                    return;
+                }
+
+                _logger.LogDebug($"Starting worker concurrency monitoring. Options: {_concurrencyOptions.Value.Format()}");
+                _timer = new System.Timers.Timer()
+                {
+                    AutoReset = false,
+                    Interval = _concurrencyOptions.Value.CheckInterval.TotalMilliseconds,
+                };
+
+                _timer.Elapsed += OnTimer;
+                _timer.Start();
+            }
+            else
+            {
+                _logger.LogDebug($"Language worker concurrency is disabled.");
+            }
+
+            await Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            if (_concurrencyOptions.Value.Enabled && _timer != null)
+            {
+                _logger.LogDebug("Stoping language worker concurrency monitoring.");
+                _timer.Stop();
+            }
+            return Task.CompletedTask;
+        }
+
+        internal async void OnTimer(object sender, System.Timers.ElapsedEventArgs e)
+        {
+            if (_disposed && _functionInvocationDispatcher == null)
+            {
+                return;
+            }
+
+            try
+            {
+                var workerStatuses = await _functionInvocationDispatcher.GetWorkerStatusesAsync();
+
+                if (AddWorkerIfNeeded(workerStatuses, _addWorkerStopWatch.Elapsed))
+                {
+                    await _functionInvocationDispatcher.StartWorkerChannel();
+                    _logger.LogDebug("New worker is added.");
+                    _addWorkerStopWatch.Restart();
+                }
+            }
+            catch (Exception ex)
+            {
+                // don't allow background exceptions to escape
+                _logger.LogError(ex.ToString());
+            }
+            _timer.Start();
+        }
+
+        internal bool AddWorkerIfNeeded(IDictionary<string, WorkerStatus> workerStatuses, TimeSpan elaspsedFromLastAdding)
```

Comment:
This method doesn't add a worker. Either rename to something like NewWorkerIsRequired, or move all the logic above into this method so it does both the check and the addition.